File: app_java/src/main/java/vlab/android/architecture/feature/home/HomeFragment.java
package vlab.android.architecture.feature.home;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.TextView;

import vlab.android.architecture.R;
import vlab.android.architecture.base.BaseFragment;
import vlab.android.architecture.feature.home.viewmodel.HomeViewModel;
import vlab.android.architecture.feature.user_repository.UserRepositoryFragment;

public class HomeFragment extends BaseFragment {

    public static final String ARG_USER_NAME = "ARG_USER_NAME";
    private OnHomeFragmentListener mListener;

    private HomeViewModel mViewModel;

    public HomeFragment() {
        // Required empty public constructor
    }

    public static HomeFragment newInstance() {
        HomeFragment fragment = new HomeFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnHomeFragmentListener) {
            mListener = (OnHomeFragmentListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.fragment_home;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    protected void initViewModel() {
        mViewModel = provideViewModel(HomeViewModel.class);

        if (getArguments() != null) {
            // TODO map argument to ViewModel
            String userName = getArguments().getString(ARG_USER_NAME);
            mViewModel.setUserName(userName);
        }
    }

    @Override
    protected void bindViewModel() {

    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ((TextView)view.findViewById(R.id.tv_welcome_msg)).setText(mViewModel.getUserName());
        if(mViewModel.isUserAuthenticated()){
            replaceFragment(R.id.content_container, new UserRepositoryFragment(), false);
        }

    }

    public void setOnHomeFragmentListener(OnHomeFragmentListener listener) {
        mListener = listener;
    }

    public interface OnHomeFragmentListener {
        // TODO: Update argument type and name
        void onHomeFragmentCallback();
    }
}

File: app_kotlin/src/main/java/vlab/android/architecture/feature/home/HomeFragment.kt
package vlab.android.architecture.feature.home

import android.content.Context
import android.os.Bundle
import android.view.View
import android.widget.TextView
import org.koin.android.viewmodel.ext.android.viewModel

import vlab.android.architecture.R
import vlab.android.architecture.base.BaseFragment
import vlab.android.architecture.feature.home.viewmodel.HomeViewModel

class HomeFragment : BaseFragment() {
    private var mListener: OnHomeFragmentListener? = null

    val mViewModel: HomeViewModel by viewModel()

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnHomeFragmentListener) {
            mListener = context
        } else {
            throw RuntimeException(context!!.toString() + " must implement OnFragmentInteractionListener")
        }
    }

    override fun getLayoutRes() = R.layout.fragment_home

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun initViewModel() {
        if (arguments != null) {
            // TODO map argument to ViewModel
            val userName = arguments!!.getString(ARG_USER_NAME)
            mViewModel!!.userName = userName
        }
    }

    override fun bindViewModel() {

    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        (view.findViewById<View>(R.id.tv_welcome_msg) as TextView).text = mViewModel!!.userName
    }

    fun setOnHomeFragmentListener(listener: OnHomeFragmentListener) {
        mListener = listener
    }

    interface OnHomeFragmentListener {
        // TODO: Update argument type and name
        fun onHomeFragmentCallback()
    }

    companion object {

        val ARG_USER_NAME = "ARG_USER_NAME"

        fun newInstance(): HomeFragment {
            val fragment = HomeFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}
